routes/api/project/location/vm.rb

```diff
@@ -5,49 +5,115 @@ class CloverApi
     @serializer = Serializers::Api::Vm
 
     r.get true do
-      vms = @project.vms_dataset.where(location: @location).authorized(@current_user.id, "Vm:view").eager(:semaphores).all
+      result = @project.vms_dataset.where(location: @location).authorized(@current_user.id, "Vm:view").paginated_result(
+        r.params["cursor"],
+        r.params["page_size"],
+        r.params["order_column"]
+      )
 
-      serialize(vms)
+      {
+        values: serialize(result[:records]),
+        next_cursor: result[:next_cursor],
+        count: result[:count]
+      }
     end
 
-    r.post true do
-      Authorization.authorize(@current_user.id, "Vm:create", @project.id)
-      fail Validation::ValidationFailed.new({billing_info: "Project doesn't have valid billing information"}) unless @project.has_valid_payment_method?
-
-      st = Prog::Vm::Nexus.assemble(
-        r.params["public_key"],
-        @project.id,
-        name: r.params["name"],
-        unix_user: r.params["unix_user"],
-        size: r.params["size"],
-        location: @location,
-        boot_image: r.params["boot_image"],
-        enable_ip4: !!r.params["enable_ip4"]
-      )
-
-      serialize(st.subject)
+    r.on "ubid" do
+      r.on String do |vm_ubid|
+        vm = Vm.from_ubid(vm_ubid)
+        handle_vm_requests(@current_user, vm)
+      end
     end
 
-    r.is String do |vm_name|
-      vm = @project.vms_dataset.where(location: @location).where { {Sequel[:vm][:name] => vm_name} }.first
+    r.on String do |vm_name|
+      r.post true do
+        Authorization.authorize(@current_user.id, "Vm:create", @project.id)
+        fail Validation::ValidationFailed.new({billing_info: "Project doesn't have valid billing information"}) unless @project.has_valid_payment_method?
+
+        required_parameters = ["public_key"]
+        allowed_optional_parameters = ["size", "unix_user", "boot_image", "enable_ip4", "private_subnet_id"]
+
+        request_body_params = Validation.validate_request_body(r.body.read, required_parameters, allowed_optional_parameters)
 
-      unless vm
-        response.status = 404
-        r.halt
+        if request_body_params["private_subnet_id"]
+          ps_id = UBID.parse(request_body_params["private_subnet_id"]).to_uuid
+          Authorization.authorize(@current_user.id, "PrivateSubnet:view", ps_id)
+          request_body_params.delete("private_subnet_id")
```

Comment:
nit: then you can remove `private_subnet_id: ps_id,` line
```suggestion
          request_body_params["private_subnet_id"] = ps_id
```